### ISSUE TITLE
id = "null" in all *ToOne(fetch=LAZY) relations when using field access type

### ISSUE THREAD

**Issue by belovaf** [yahoo/elide#844]
## Expected Behavior
Fetched relations always have correctly loaded id.

## Current Behavior
id = "null" in all *ToOne(fetch=LAZY) relations when using field access type.
For example Elide response is:
```
{
  "data": {
    "type": "parentEntity",
    "id": "123",
    "attributes": {
      "attr1": 1
    },
    "relationships": {
      "childEntityAttr": {
        "data": {
          "type": "childEntity",
          "id": "null"
        }
      }
    }
  }
}
```
When fetch=LAZY, hibernate creates proxy object. Hibernate doesn't fill proxy's fields, but delegates all methods to real object. So you can't simply read proxy's fields.

## Possible Solution
Never use ```accessType = AccessType.FIELD``` in ```com.yahoo.elide.core.EntityBinding```.
Regardless of Hibernate accessType we should use property access.

## Steps to Reproduce (for bugs)
1. Create entities: Parent <-- OneToOne(fetch=LAZY) --> Child
2. Fetch parent
3. Fetched child relation will have id = "null"

## Context
Try to apply hibernate annotations to Entity fields.

## Your Environment
* Elide version used: 4.3.0
* Environment name and version (Java 1.8.0_152): Java 1.8
* Operating System and version: any

**Comment by belovaf** [comment]
I can make PR if you accept my possible solution.

**Comment by aklish** [comment]
`Never use accessType = AccessType.FIELD in com.yahoo.elide.core.EntityBinding` seems a bit strong.  

Maybe only for fetching the ID field, and only if the property accessor exists?

I'd be happy to review your PR.

**Comment by belovaf** [comment]
In my opinion, we should not bypass encapsulation and read object fields directly. Property access could add mach more flexibility.
What about read from fields directly only if the property accessor doesn't exists?

**Comment by aklish** [comment]
Field level accessors was one of our more requested features - and is used in production by a number of projects.  We cannot cripple that capability.   These projects will not want to add methods for their entities (for readability).

As far as encapsulation - I generally agree with you.  However, this is a standard approach for JPA, so it has already been embraced by the community. 

Maybe there is another mechanism to tell the ORM to hydrate the lazy field?

**Comment by belovaf** [comment]
Sorry for my very poor English. :)
I try to explain my opinion and possible solution.
1) Writing to fields directly is absolutely ok for jpa provider because it hydrates state of the entity.
But jpa returns to you encapsulated object. You shouldn't write/read the object fields directly.
2) We can still use field access, no problem. But i think, that property accessor should have greater priority.
3) You can check entity for provider specific proxy class. If entity is proxy, you can get the delegated instance inside proxy and read it fields, but it seems hacky and not solid. :)

Surely, concrete decision is up to you, i can implement any solution, just want to explain my point of view.

**Comment by wcekan** [comment]
`lazy` with `OneToOne` and `optional=true` never quite works even with property accessors.

However, you should never mix field and property in a single bean.  Hibernate assumes `AccessType.FIELD` whenever the `@Id` is on the field, regardless of existence of getter/setter.  The proxied FIELD access bean should always hydrate the fields, but a PR with a test that exposes the problem would be useful.

https://docs.jboss.org/hibernate/orm/5.1/userguide/html_single/chapters/domain/access.html

**Comment by aklish** [comment]
I agree with @wcekan that maybe the first step is to submit a small PR with a test that breaks.  Then we can evaluate solutions.
